pkg/httpproxy/sign.go

```diff
@@ -0,0 +1,230 @@
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"github.com/rancher/norman/httperror"
+	"github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/rancher/types/apis/core/v1"
+)
+
+type Signer interface {
+	sign(*http.Request, v1.SecretInterface, string) error
+}
+
+func newSigner(auth string) Signer {
+	splitAuth := strings.Split(auth, " ")
+	switch strings.ToLower(strings.ToLower(splitAuth[0])) {
+	case "awsv4":
+		return awsv4{}
+	case "bearer":
+		return bearer{}
+	case "basic":
+		return basic{}
+	case "digest":
+		return digest{}
+	case "arbitrary":
+		return arbitrary{}
+	}
+	return nil
+}
+
+func (br bearer) sign(req *http.Request, secrets v1.SecretInterface, auth string) error {
+	data, secret, err := getAuthData(auth, secrets, []string{"passwordField", "credID"})
+	if err != nil {
+		return err
+	}
+	req.Header.Set(AuthHeader, fmt.Sprintf("%s %s", "Bearer", secret[data["passwordField"]]))
+	return nil
+}
+
+func (b basic) sign(req *http.Request, secrets v1.SecretInterface, auth string) error {
+	data, secret, err := getAuthData(auth, secrets, []string{"usernameField", "passwordField", "credID"})
+	if err != nil {
+		return err
+	}
+	key := fmt.Sprintf("%s:%s", secret[data["usernameField"]], secret[data["passwordField"]])
+	encoded := base64.URLEncoding.EncodeToString([]byte(key))
+	req.Header.Set(AuthHeader, fmt.Sprintf("%s %s", "Basic", encoded))
+	return nil
+}
+
+func (a awsv4) sign(req *http.Request, secrets v1.SecretInterface, auth string) error {
+	_, secret, err := getAuthData(auth, secrets, []string{"credID"})
+	if err != nil {
+		return err
+	}
+	service, region := a.getServiceAndRegion(req.URL.Host)
+	creds := credentials.NewStaticCredentials(secret["accessKey"], secret["secretKey"], "")
+	awsSigner := v4.NewSigner(creds)
+	var body []byte
+	if req.Body != nil {
+		body, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			return fmt.Errorf("error reading request body %v", err)
+		}
+	}
+	_, err = awsSigner.Sign(req, bytes.NewReader(body), service, region, time.Now())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a awsv4) getServiceAndRegion(host string) (string, string) {
+	//format : service.region.amazonaws.com
+	parts := strings.Split(host, ".")
+	if len(parts) != 4 {
+		return "", ""
+	}
+	return parts[0], parts[1]
+}
+
+func (d digest) sign(req *http.Request, secrets v1.SecretInterface, auth string) error {
+	data, secret, err := getAuthData(auth, secrets, []string{"usernameField", "passwordField", "credID"})
+	if err != nil {
+		return err
+	}
+	resp, err := doNewRequest(req) // request to get challenge fields from server
+	if err != nil {
+		return err
+	}
+	challengeData, err := parseChallenge(resp.Header.Get("WWW-Authenticate"))
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("digest challenge data %v", challengeData)
```

Comment:
Hope we dont log any sensitive info here

Reply:
it's the details (nonce,  etc) server responds which we then use for authenticate request, this usually should be new each time but I'll remove it to be on safer side